srtcore/core.cpp

```diff
@@ -14468,3 +14531,1224 @@ int CUDTGroup::configure(const char* str)
+        {
+            // Found you, buddy. Go on.
+            return this_link;
+        }
+
+        if (this_link == state.ilink)
+        {
+            // No more links. Sorry.
+            return gli_NULL();
+        }
+
+        // Check maybe next link...
+    }
+
+    return this_link;
+}
+
+struct LinkCapableData
+{
+    CUDTGroup::gli_t link;
+    int flight;
+};
+
+CUDTGroup::gli_t CUDTGroup::linkSelect_window(const CUDTGroup::BalancingLinkState& state)
+{
+    if (state.ilink == gli_NULL())
+    {
+        // Very first sending operation. Pick up the first link
+        return m_Group.begin();
+    }
+
+
+    gli_t this_link = gli_NULL();
+
+    if (m_RandomCredit <= 0)
+    {
+        vector<LinkCapableData> linkdata;
+        int total_flight = 0;
+        int number_links = 0;
+
+        // First, collect data required for selection
+        vector<gli_t> linkorder;
+
+        gli_t last = state.ilink;
+        ++last;
+        // NOTE: ++last could make it == m_Group.end() in which
+        // case the first loop will get 0 passes and the second
+        // one will be from begin() to end().
+        for (gli_t li = last; li != m_Group.end(); ++li)
+            linkorder.push_back(li);
+        for (gli_t li = m_Group.begin(); li != last; ++li)
+            linkorder.push_back(li);
+
+        // Sanity check
+        if (linkorder.empty())
+        {
+            LOGC(dlog.Error, log << "linkSelect_window: IPE: no links???");
+            return gli_NULL();
+        }
+
+        // Fallback
+        this_link = *linkorder.begin();
+
+        // This does the following:
+        // We have links: [ 1 2 3 4 5 ]
+        // Last used link was 4
+        // linkorder: [ (5) (1) (2) (3) (4) ]
+        for (vector<gli_t>::iterator i = linkorder.begin(); i != linkorder.end(); ++i)
+        {
+            gli_t li = *i;
+            int flight = li->ps->core().m_iSndMinFlightSpan;
+
+            HLOGC(dlog.Debug, log << "linkSelect_window: previous link was #" << distance(m_Group.begin(), state.ilink)
+                    << " Checking link #" << distance(m_Group.begin(), li)
+                    << "@" << li->id << " TO " << SockaddrToString(li->peer)
+                    << " flight=" << flight);
+
+            // Upgrade idle to running
+            if (li->sndstate == GST_IDLE)
+                li->sndstate = GST_RUNNING;
+
+            if (li->sndstate != GST_RUNNING)
+            {
+                HLOGC(dlog.Debug, log << "linkSelect_window: ... state=" << StateStr(li->sndstate) << " - skipping");
+                // Skip pending/broken links
+                continue;
+            }
+
+            // Check if this link was used at least once so far.
+            // If not, select it immediately.
+            if (li->load_factor == 0)
+            {
+                HLOGC(dlog.Debug, log << "linkSelect_window: ... load factor empty: SELECTING.");
+                this_link = li;
+                goto ReportLink;
```

Comment:
Avoid goto labels. Make a function `CUDTGroup::gli_t reportLink(CUDTGroup::gli_t this_link)` and return its result:
```
return reportLink(this_link);
```

Reply:
Good idea.